src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -512,33 +523,112 @@ private static ItemGroup getItemGroup(Run<?, ?> build) {
-                cnt = list.length;
+            targetDir.mkdirs();  // Create target if needed
+            List<String> list = scan(srcDir, expandedFilter, expandedExcludes);
+            for (String file : list) {
+                copyOne(src, dst, fingerprints, srcDir.child(file), new FilePath(targetDir, isFlatten() ? file.replaceFirst(".+/", "") : file), md5);
             }
-
+            int cnt = list.size();
             console.println(Messages.CopyArtifact_Copied(cnt, HyperlinkNote.encodeTo('/'+ src.getParent().getUrl(), src.getParent().getFullDisplayName()),
                     HyperlinkNote.encodeTo('/'+src.getUrl(), Integer.toString(src.getNumber()))));
             // Fail build if 0 files copied unless copy is optional
             return cnt > 0 || isOptional();
         } finally {
-            copier.end();
+            if (fingerprints != null) {
+                for (Run<?, ?> r : new Run<?, ?>[] {src, dst}) {
+                    if (fingerprints.size() > 0) {
+                        Fingerprinter.FingerprintAction fa = r.getAction(Fingerprinter.FingerprintAction.class);
+                        if (fa != null) {
+                            fa.add(fingerprints);
+                        } else {
+                            r.addAction(new Fingerprinter.FingerprintAction(r, fingerprints));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static List<String> scan(VirtualFile root, String expandedFilter, @CheckForNull String expandedExcludes) throws IOException {
+        List<String> r = new ArrayList<>();
+        // TODO need VirtualFile.list(String, String, boolean) like FilePath offers
+        List<TokenizedPattern> patts = new ArrayList<>();
+        if (expandedExcludes != null) {
+            for (String patt : expandedExcludes.split(",")) {
+                patts.add(new TokenizedPattern(normalizePattern(patt)));
+            }
+        }
+        FILE: for (String child : root.list(expandedFilter)) {
+            child = child.replace('\\', '/'); // TODO list(String) ought to specify `/` as the separator
+            for (TokenizedPattern patt : patts) {
+                if (patt.matchPath(new TokenizedPath(child), true)) {
+                    continue FILE;
+                }
+            }
+            r.add(child);
+        }
+        return r;
+    }
+
+    /** Similar to a method in {@link DirectoryScanner}. */
+    private static String normalizePattern(String p) {
+        String pattern = p.replace('\\', '/'); // we only deal with forward slashes here
+        if (pattern.endsWith("/")) {
+            pattern += SelectorUtils.DEEP_TREE_MATCH;
+        }
+        return pattern;
+    }
+
+    private static void copyOne(Run<?,?> src, Run<?,?> dst, Map<String, String> fingerprints, VirtualFile s, FilePath d, MessageDigest md5) throws IOException, InterruptedException {
```

Comment:
same, does not seem to be plugin-specific

Reply:
It is pretty specific to `copyartifact` I think.